Assignment 2/src/Vault.sol

```diff
@@ -0,0 +1,99 @@
+///
+///     Amounts are stored with the same decimals as the token contract that
+///     was used to initialize the `Vault`.
+contract Vault {
+    /// @notice The token that is stored in the vault.
+    IERC20 immutable public token;
+
+    /// @notice This event is emitted when a successful deposit is made. It
+    ///     contains the account that made the deposit and the amount that was
+    ///     deposited.
+    event Deposit(address indexed account, uint256 deposited);
+
+    /// @notice This event is emitted when a successful withdrawal is
+    ///     performed. The first argument is the account that made the
+    ///     withdrawal and the second argument is the amount that was
+    ///     withdrawn.
+    event Withdrawal(address indexed account, uint256 withdrawn);
+
+    /// @notice Deposits made using `deposit` can be reverted with this error
+    ///     if the transfer failed.
+    /// @dev This error is thrown if `transferFrom` returns `false`. Instead of
+    ///     returning false however, it is possible that the `IERC20` reverts,
+    ///     for example when attempting to transfer more than the allowance.
+    ///     So: not all failed deposits will throw this error.
+    error DepositFailed(address account, uint256 amount);
+    /// @notice Withdrawals made using `withdraw` can be reverted with this
+    ///     error if the amount is too low or if the transfer fails for
+    ///     whatever reason. Like with deposits, it is possible that transfers
+    ///     revert before this error can be thrown. See also `DepositFailed`.
+    error WithdrawalFailed(address account, uint256 amount);
+
+    /// @notice The balance, the size of the deposit for an address.
+    mapping(address => uint256) public depositOf;
+
+    /// @notice Deploy the contract.
+    /// @param _token The token that should be accepted in the vault.
+    constructor(IERC20 _token) {
+        token = _token;
+    }
+
+    /// @notice Deposit tokens into the vault and add them to the vault
+    ///     balance. Make sure that this contract is allowed to transfer the
+    ///     tokens.
+    ///
+    //      This function may revert, for example by throwing an
+    ///     `DepositFailed`. However, it may fail if the token transfer fails.
+    ///     See also the documentation for `DepositFailed`.
+    ///
+    /// @param amount The amount of tokens to transfer.
+    /// @dev Follows the Checks-Effects-Interactions Pattern.
+    function deposit(uint256 amount) external {
+        // Add to the deposit
+        depositOf[msg.sender] += amount;
+        // Transfer the tokens to this contract
+        bool successful = token.transferFrom(msg.sender, address(this), amount);
+
+        // Finally, check if the transfer was successful to either revert with
+        // an error or emit an event.
+        if (successful) {
+            emit Deposit(msg.sender, amount);
+        } else {
+            revert DepositFailed(msg.sender, amount);
```

Comment:
its definitely a graceful way to handle errors.  i usually just let it revert lol